lib/tapioca/commands/abstract_gem.rb

```diff
@@ -67,12 +77,98 @@ def initialize(
+        @include_annotations = include_annotations
         @include_exported_rbis = include_exported_rbis
+        @annotation_sources = annotation_sources
+        @annotation_auth = annotation_auth
+        @annotation_netrc_file = annotation_netrc_file
+        @annotation_netrc_info = T.let(nil, T.nilable(Netrc))
+        @excluded_annotations = excluded_annotations
+        @tokens = T.let(repo_tokens, T::Hash[String, T.nilable(String)])
+        @indexes = T.let({}, T::Hash[String, RepoIndex])
         @halt_upon_load_error = halt_upon_load_error
       end
 
       private
 
+      sig { params(repo_uri: String, path: String).returns(T.nilable(String)) }
+      def fetch_file(repo_uri, path)
+        if repo_uri.start_with?(%r{https?://})
+          fetch_http_file(repo_uri, path)
+        else
+          fetch_local_file(repo_uri, path)
+        end
+      end
+
+      sig { params(repo_uri: String, path: String).returns(T.nilable(String)) }
+      def fetch_local_file(repo_uri, path)
+        File.read("#{repo_uri}/#{path}")
+      rescue => e
+        say_error("\nCan't fetch file `#{path}` (#{e.message})", :bold, :red)
+        nil
+      end
+
+      sig { params(repo_uri: String, path: String).returns(T.nilable(String)) }
+      def fetch_http_file(repo_uri, path)
+        auth = @tokens[repo_uri]
+        uri = URI("#{repo_uri}/#{path}")
+
+        request = Net::HTTP::Get.new(uri)
+        request["Authorization"] = auth if auth
+
+        response = Net::HTTP.start(uri.hostname, uri.port, use_ssl: uri.scheme == "https") do |http|
+          http.request(request)
+        end
+
+        case response
+        when Net::HTTPSuccess
+          response.body
+        else
+          say_http_error(path, repo_uri, message: response.class)
+          nil
+        end
+      rescue SocketError, Errno::ECONNREFUSED => e
+        say_http_error(path, repo_uri, message: e.message)
+        nil
+      end
+
+      sig { params(repo_uri: String, repo_number: T.nilable(Integer)).returns(T.nilable(RepoIndex)) }
+      def fetch_annotation_source_index(repo_uri, repo_number:)
+        say("Retrieving index from central repository#{repo_number ? " ##{repo_number}" : ""}... ", [:blue, :bold])
+        content = fetch_file(repo_uri, CENTRAL_REPO_INDEX_PATH)
+        return unless content
+
+        index = RepoIndex.from_json(content)
+        say("Done", :green)
+        index
+      end
+
+      sig { returns(T::Hash[String, RepoIndex]) }
+      def fetch_annotation_source_indexes
+        multiple_repos = @annotation_sources.size > 1
+        repo_number = 1
+        indexes = T.let({}, T::Hash[String, RepoIndex])
+
+        @annotation_sources.each do |uri|
+          index = fetch_annotation_source_index(uri, repo_number: multiple_repos ? repo_number : nil)
+          next unless index
+
+          indexes[uri] = index
+          repo_number += 1
+        end
+
+        if indexes.empty?
+          raise Thor::Error, set_color(
```

Comment:
I think we have `say_error` for formatting?

Reply:
I think this is more about control flow. There are two other cases of this pattern in the `abstract_gem` file.